lib/smf/smf.c

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright 2021 The Chromium OS Authors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "smf.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(smf);
+
+/*
+ * Private structure (to this file) used to track state machine context.
+ * The structure is not used directly, but instead to cast the "internal"
+ * member of the smf_ctx structure.
+ */
+struct internal_ctx {
+	const struct smf_state *last_entered;
+	uint32_t running : 1;
+	uint32_t enter	 : 1;
+	uint32_t exit	 : 1;
+};
+
+#ifndef CONFIG_SMF_FLAT
+/* Gets the first shared parent state between a and b (inclusive) */
+static const struct smf_state *shared_parent_state(const struct smf_state *a,
+				const struct smf_state *b)
+{
+	const struct smf_state *orig_b = b;
+
+	/* There are no common ancestors */
+	if (b == NULL)
+		return NULL;
+
+	/* This assumes that both A and B are NULL terminated without cycles */
+	while (a != NULL) {
+		/* We found a match return */
+		if (a == b)
+			return a;
+
+		/*
+		 * Otherwise, increment b down the list for comparison until we
+		 * run out, then increment a and start over on b for comparison
+		 */
+		if (b->parent == NULL) {
+			a = a->parent;
+			b = orig_b;
+		} else {
+			b = b->parent;
+		}
+	}
+
+	return NULL;
+}
+#endif
+
+/*
+ * Call all entry functions of parents before children. If set_state is called
+ * during one of the entry functions, then do not call any remaining entry
+ * functions.
+ */
+static void call_entry_functions(struct smf_ctx *const ctx,
+					struct internal_ctx *const internal,
+			       const struct smf_state *stop,
+			       const struct smf_state *current)
+{
+	if (current == stop) {
+		return;
+	}
+
+#ifndef CONFIG_SMF_FLAT
+	call_entry_functions(ctx, internal, stop, current->parent);
+#else
+	call_entry_functions(ctx, internal, stop, NULL);
+#endif
+
+	/*
+	 * If the previous entry function called set_state, then don't enter
+	 * remaining states.
+	 */
+	if (!internal->enter) {
+		return;
+	}
+
+	/* Track the latest state that was entered, so we can exit properly. */
+	internal->last_entered = current;
+	if (current->entry) {
+#ifdef CONFIG_SMF_PRINT_MSG
+		if (current->entry_msg) {
+			printk("%s\n", current->entry_msg);
```

Comment:
it's better to use logging subsystem for that.

Reply:
This has been removed in favor of the logging system